coremasscheck.F90

```diff
@@ -0,0 +1,66 @@
+! Include shortname defintions, so that the F77 code does not have to be modified to
+! reference the CARMA structure.
+#include "carma_globaer.h"
+
+!!  This routine checks if the coremass exceeds the total
+!!
+!! @author Yunqian Zhu
+!! @version Apr-2021
+subroutine coremasscheck(carma, cstate, iz, fixcoremass,logmsg,abort, packagename, rc)
+
+  ! types
+  use carma_precision_mod
+  use carma_enums_mod
+  use carma_constants_mod
+  use carma_types_mod
+  use carmastate_mod
+  use carma_mod
+
+  implicit none
+
+  type(carma_type), intent(inout)      :: carma       !! the carma object
+  type(carmastate_type), intent(inout) :: cstate      !! the carma state object
+  integer, intent(in)                  :: iz
+  logical, intent(in)                  :: fixcoremass !! if we fix the coremass if exceeds total
+  logical, intent(in)                  :: logmsg      !! write message to log file if core mass check fails
+  logical, intent(in)                  :: abort       !! set return code to ERROR and return if  if core mass check fails
+  character(len=*), intent(in)         :: packagename !! string to add to error message
+  integer, intent(inout)               :: rc          !! return code, negative
+
+  ! Local declarations
+  integer                        :: igroup,ibin
+  integer                        :: iepart,i,icore
+  real(kind=f)                   :: coremass
+  real(kind=f)                   :: factor
+
+  real(kind=f), parameter :: thresh  = 1e-10_f     ! relative threshold for core mass and kappa checking
+  real(kind=f), parameter :: thresh1 = 1._f+thresh
+
+  ! check the coremass exceeding the total mass
+  do igroup = 1,NGROUP
+     iepart = ienconc(igroup)
+     do ibin = 1,NBIN
+        if (pc(iz, ibin, iepart) > 0._f) then
+           coremass = 0._f
+           do i = 1, ncore(igroup)
+              icore = icorelem(i,igroup)
+              coremass = coremass + pc(iz, ibin, icore)
+           end do ! i = 1, ncore(igroup)
+           if (coremass > pc(iz, ibin, iepart) * rmass(ibin, igroup) * thresh1) then
+              if(fixcoremass)then
+                 pc(iz, ibin, iepart) = coremass/rmass(ibin,igroup) + SMALL_PC
```

Comment:
Why use two different constants? If using a threhold I think you should make sure that the results is bigger than a threshold not smallpc.

Reply:
@zywshoon

Reply:
@cbardeen this one is to add some small values to total mass so it is larger than coremass, so I just add a very small value, not the round-up value.

Reply:
is done (code change by Yunqian)

Reply:
What the code is showing me is that it is still SMALL_PC and not the threshold fraction. Doesn't seem done to me.

Reply:
Chuck, it looks like to me in the code that , if coremass > the total mass
and 
1. if core mass smaller the total mass plus a round off error then we increase the total mass while adding a small value, this is done in any case.
2. if coremass however is larger the total mass and larger than the total mass plus the round-off error, then either we fix this with adding a small value to the total mass, or we print and error (depending on the setting).

I think, this is what we had discussed, the small value is added as Yunqian has described on August 23rd. It that not correct?

Reply:
We agreed to change the SMALL_PC value to "almost_zero" value. SMALL_PC could be removed altogether, however, Yunqian is in favor or adding a very tiny value to have sulfate, not zero, which may be important for adding the nitrate model later.

Reply:
Replace SMALL_PC with ALMOST_ZERO

Reply:
ALMOST_ZERO is now used